0012-common-ledger-api/0012-common-ledger-api.md

```diff
@@ -0,0 +1,1201 @@
+# Common Ledger API
+
+The Common Ledger API is a RESTful API served by a ledger (or an adapter), which provides functionality necessary for ILP compatibility. The Common Ledger API provides a single standard API that a ledger can serve in order to ease integration with other Interledger Protocol components and applications, such as the reference ILP Client and ILP Connector. This is not the only way a ledger can become ILP-enabled, but it provides a template that minimizes the integration work necessary for compatibility with ILP software.
+
+## Overview
+
+The Common Ledger API defines the following data types and structures:
+
+- [Transfer resource][]
+- [Account resource][]
+- [Message resource][]
+- [Crypto-Condition][] and [Crypto-Condition Fulfillment][]
+- [Error Codes](#api-error-codes)
+
+The core operations of the API are:
+
+- [Get Ledger Metadata][]
+- [Prepare Transfer][]
+    - If the transfer is unconditional, it executes immediately.
+    - If the transfer is conditional, it waits for a matching fulfillment.
+- [Submit Fulfillment][]
+    - Executes a conditional transfer, if the fulfillment matches the condition.
+- [Reject Transfer][]
+- [Get Transfer][] and check its status
+- [Get Transfer Fulfillment][]
+- [Get Account][] info including balance
+- [Send Message][] to another account
+- [Get Auth Token][] for [Authorization and Authentication](#authorization-and-authentication)
+- [Websocket][] sub-API for subscribing to transfers and accounts
+
+The ledger MAY support additional methods, including but not limited to:
+
+- Create or Update Account
+- Health check status
+- Look up transfers by account or [Crypto-Condition][]
+
+
+## Conventions
+
+The API is designed to be [RESTful](https://en.wikipedia.org/wiki/Representational_state_transfer), extensible, and modular. By design, every resource type contains links to other methods, so you can "naturally" discover other API paths by starting from any object. Clients SHOULD NOT hardcode any [URL][] other than the path to the [Get Ledger Metadata][] method. A ledger MAY host different objects and API methods at different hostnames or base URLs, as long as all client applications can dereference the URLs in the usual way.
```

Comment:
I'm wondering whether gPRC will be a better approach in the long term. There are many benefits out of the box. Transparent binary serialization, cross-language support, interface (auto)documentation in .proto files, better performance. No need for "parallel" websockets to transmit async events.